File: lib/utils/getMinDistBetweenEnteringPoints.ts
import { NodeWithPortPoints } from "lib/types/high-density-types"
import { getPortPointsFromNodeWithPortPoints } from "./getPortPointsFromNodeWithPortPoints"

export const getMinDistBetweenEnteringPoints = (node: NodeWithPortPoints) => {
  let minDist = Infinity
  const points = getPortPointsFromNodeWithPortPoints(node)

  // Compare each point with every other point
  for (let i = 0; i < points.length; i++) {
    for (let j = i + 1; j < points.length; j++) {
      const p1 = points[i]
      const p2 = points[j]

      // Skip points on different layers
      if (p1.z !== p2.z) {
        continue
      }

      // Skip points from the same root net (they can share junction points)
      if (
        p1.rootConnectionName &&
        p1.rootConnectionName === p2.rootConnectionName
      ) {
        continue
      }

      // Calculate Euclidean distance between points
      const dist = Math.sqrt((p1.x - p2.x) ** 2 + (p1.y - p2.y) ** 2)

      minDist = Math.min(minDist, dist)
    }
  }

  return minDist === Infinity ? 0 : minDist
}

File: rust/capacity-autorouter/src/utils/get_min_dist_between_entering_points.rs
use serde_json::Value;

pub fn get_min_dist_between_entering_points(node: &Value) -> f64 {
    let mut min_dist = f64::INFINITY;
    let points = node["portPoints"].as_array().expect("node portPoints");
    for i in 0..points.len() {
        for j in i + 1..points.len() {
            let p1 = &points[i];
            let p2 = &points[j];
            if p1.get("z") != p2.get("z") {
                continue;
            }
            if p1["rootConnectionName"]
                .as_str()
                .is_some_and(|s| !s.is_empty())
                && p1["rootConnectionName"] == p2["rootConnectionName"]
            {
                continue;
            }
            let dx =
                p1["x"].as_f64().expect("port point x") - p2["x"].as_f64().expect("port point x");
            let dy =
                p1["y"].as_f64().expect("port point y") - p2["y"].as_f64().expect("port point y");
            min_dist = min_dist.min((dx * dx + dy * dy).sqrt());
        }
    }
    if min_dist == f64::INFINITY {
        0.0
    } else {
        min_dist
    }
}
